README.md
<div align="center">
  <img alt="logo" src="https://user-images.githubusercontent.com/1403842/28923702-d8155d46-7899-11e7-817b-1193d138e5b8.png" width="192"/>
</div>

<div align="center">
  <h1>VEDA</h1>
  <i>VJ / Live Coding on Atom with GLSL.</i>
  <br>
  <br>
  <br>
  <img alt="screenshot" src="https://user-images.githubusercontent.com/1403842/28673275-1d42b062-731d-11e7-92b0-bde5ca1f1cae.gif" style="width: 100% !important;"/>
  <br>
  <br>
</div>

<div align="center">

![TravisCI](https://img.shields.io/travis/fand/veda.svg)
![apm version](https://img.shields.io/apm/v/veda.svg)
![license MIT](https://img.shields.io/apm/l/veda.svg)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![Greenkeeper badge](https://badges.greenkeeper.io/fand/veda.svg)](https://greenkeeper.io/)

</div>
<br>
<br>

##### TOC

- [Features](#features)
- [Install](#install)
- [Usage](#usage)
- [Examples](#examples)

## What's this?

`VEDA` is a GLSL runtime environment for Atom.
When you write GLSL code in Atom, VEDA immediately evaluates it and shows the result on the background.
It's just like [GLSL sandbox](http://glslsandbox.com/) or [Shadertoy](https://www.shadertoy.com/), but you can use autocomplete and linter by using existing Atom packages.
Moreover, It supports Audio inputs , MIDI inputs, loading videos and images, etc...!!!!

`VEDA` has following features.

- Fragment shaders runtime like [GLSL Sandbox](http://glslsandbox.com/)
- Vertex shader runtime like [vertexshaderart.com](https://www.vertexshaderart.com/)
- Loading images / videos
- Additional `uniform` variables useful for live coding
  - Audio input
  - MIDI input
  - OSC input
  - Webcam input
  - Keyboard input
  - Gamepad input
- Auto completion (thx to [autocomplete-glsl](https://atom.io/packages/autocomplete-glsl/))
- Linting (thx to [linter-glsl](https://atom.io/packages/linter-glsl/))

## Tutorial

- [English](https://medium.com/@amagitakayosi/vj-live-coding-on-atom-glsl-livecoder-329eec5462df)
- [日本語](http://blog.gmork.in/entry/2017/08/04/173000)

## Install

Just install from Atom GUI or `apm`.

```
$ apm install veda
```

Sometimes Atom shows an error like below.
If you see this, please try rebuilding the package from 🐞 icon on the footer.

```
Failed to require the main module of 'veda' because it requires an incompatible native module.
Run `apm rebuild` in the package directory to resolve.
```

## Features

### Commands

`VEDA` installs following commands to Atom.

- `toggle`
  - Start / Stop VEDA.
- `load-shader` (key: `ctrl-enter`)
  - Load the shader on current editor.
- `watch-shader` (key: `ctrl-shift-enter`)
  - Watch current tab and load the shader automatically.
- `watch-active-editor` (key: `ctrl-alt-enter`)
  - Watch active tab and run `watch-shader` automatically.
- `stop-watching` (key: `ctrl-.`)
  - Stop `watch-shader` and `watch-active-editor`.
- `toggle-fullscreen` (key: `ctrl-escape`)
  - Show the output fullscreen in the window

A typical workflow can be like this:

1.  Enable `VEDA` by running `veda:toggle` from the Command Palette of Atom.
2.  Edit your GLSL code.
3.  Hit `ctrl-enter` to run `veda:load-shader`.

### Preset `uniform` variables

- `float time`:
  - The elapsed time since `VEDA` has started.
- `vec2 resolution`
  - The resolution of the screen.
- `vec2 mouse`
  - Current position of mouse.
  - `vec2(0)` to `vec2(1)`
- `sampler2D backbuffer`
  - Rendered result of last frame.
  - RGBA format
- `sampler2D samples`
  - Samples of the audio input.
- `sampler2D spectrum`
  - FFT result of the audio input.
- `float volume`
  - The volume of the audio input.
- `sampler2D midi`
  - Last MIDI event for each channel of MIDI devices.
  - `x`: 3rd byte of the event
- `sampler2D note`
  - States of note numbers of MIDI devices.
  - `x`: the volume of the note

### Settings

The settings of `VEDA` can be configured in 3 ways: global settings, project settings, and file settings.

- Global settings are loaded from Settings page of Atom.
- Project settings are loaded from `.vedarc`.
- File settings are loaded from the comments of the shader.

The order of priority is as follows:

`File Settings > Project Settings > Global Settings`

When File Settings and Global Settings has same properties, File Settings are used.

#### Global Settings

Global settings are most general settings.
You can change settings in `Settings` page of Atom.

If there are no project `.vedarc` or valid comments, VEDA will use the global settings as default.

#### Project Settings: `.vedarc`

Project settings is loaded from `.vedarc` on your project root.

- `.vedarc` must be located in your project's root directory.
- `.vedarc` is parsed as [JSON5 format](https://github.com/json5/json5).
  - You can write comments in `.vedarc`.
- `.vedarc` is loaded on startup and reloaded automatically when you edit it.

For example, when you write `.vedarc` like this:

```javascript
{
  "IMPORTED": {
    "image1": {
      "PATH": "./1.jpg",
    },
  },
  "vertexMode": "LINES",
  "pixelRatio": 2,
  "audio": true,
  "midi": true,
}
```

Then `VEDA` interpret like this:

- Load `./1.jpg` as a texture `image1`
- Draw lines on vertex shaders
- Enable audio input
- Enable MIDI input

#### File Settings

You can also write settings specific for the file.
Write comments on the head of the file like this:

```glsl
/*{ "audio": true }*/

void main() {
    ...
}
```

The comment must be written in the same format as `.vedarc`.

## Examples

### Fragment Shaders

You can write fragment shaders like [GLSL Sandbox](http://glslsandbox.com).

Fragment shaders must be named like `*.frag`.
Create a file `foo.frag` like this:

```glsl
precision mediump float;
uniform float time;
uniform vec2 resolution;

void main() {
    vec2 uv = gl_FragCoord.xy / resolution.xy;
    gl_FragColor = vec4(uv, 0.5 + 0.5 * sin(time), 1.0);
}
```

Then save it and hit `ctrl-enter` to run it.
VEDA will show the result on the background.

See [examples](./examples/shader1.frag) for actual usage.

### Vertex Shaders

VEDA also supports vertex shaders like [vertexshaderart.com](https://vertexshaderart.com/).

Vertex shaders must be named like `*.vert`.
Create a file `foo.vert` like this:

```glsl
/*{ "vertexCount": 300 }*/
precision mediump float;
attribute float vertexId;
uniform float vertexCount;
uniform float time;
uniform vec2 resolution;
varying vec4 v_color;

void main() {
  float i = vertexId + time *2.;

  vec3 pos = vec3(
    cos(i * 1.0),
    sin(i * 1.1),
    cos(i * 1.2)
  );

  gl_Position = vec4(pos.x, pos.y, pos.z, 1);

  v_color = vec4(fract(vertexId / 3.), 1, 1, 1);
}
```

Then save it and hit `ctrl-enter` to run it.
VEDA will show the result on the background.

See [examples](./examples/vertex.vert) for actual usage.

### Optional Inputs

To use these features, you have to enable them by adding following lines to `.vedarc` or header comments.

- Audio inputs: `"audio": true`
- MIDI inputs: `"midi": true`
- Webcam inputs: `"camera": true`
- Keyboard inputs: `"keyboard": true`
- Gamepad inputs: `"gamepad": true`

### Audio inputs

You can use audio data of the audio input.
These data are obtained by AnalyserNode of [Web Audio API](https://webaudio.github.io/web-audio-api/).

`sampler2D samples` stores the most recent 256 frames from the audio input.
This is useful for drawing waveforms.

`sampler2D spectrum` stores the FFT result.
This is useful to draw the volume of specific frequency band, such as spectrum visualizer.

`float volume` is the average of all the frequency bands in `spectrum`.
See [examples](./examples/audio.frag) for actual usage.

### MIDI Events

`sampler2D midi` stores MIDI events obtained by [Web MIDI API](https://www.w3.org/TR/webmidi/).
The size of `midi` is `256x128`.
Each pixel stores the last event of the corresponding MIDI Events.

For example, `texture2D(midi, vec2(144. / 256., 0)).x` yields the note number of last `note on` event of MIDI Channel 1.

- `144.` (0x90): `note on` event of MIDI Channel 1
- `.x` (2nd byte): Note number

See [examples](./examples/midi.frag) for actual usage.

`sampler2D note` stores the volumes for each note number
The size of `midi` is `128x1`.
Each pixel stores the volume of the last event for corresponding MIDI note.

For example, `texture2D(note, vec2(60. / 128., 0)).x` yields the volume of note `C4` (Middle C).

See [examples](./examples/note.frag) for actual usage.

### OSC Inputs

VEDA accepts OSC messages on the port written in `osc` property of the settings.
When you write `"osc": 4000` to `.vedarc` or the header comment, messages will be stored and passed as textures:

- Texture name will be automatically generated from addresses.
  - `/foo`: `sampler2D osc_foo`
  - `/foo/bar`: `sampler2D osc_foo_bar`
- Arguments are translated to float. Strings are ignored.
  - `/foo 0.1 hello 100` yields a texture that contains `[0.1 0 100]`

See [examples](./examples/osc.frag) for actual usage.

### Webcam Inputs

`sampler2D camera` stores the images from the webcam.
`texture2D(camera, uv)` returns vec3 color.

See [examples](./examples/camera.frag) for actual usage.

### Keyboard Inputs

`sampler2D key` stores the status of keyboard.
The size of `keyboard` is `256x1`.

For example, `texture2D(key, vec2(65. / 256., 0.))` returns 1.0 when `a` is pressed.

Hitting `ESC` key resets the states of all key inputs.

See [examples](./examples/key.frag) for actual usage.

### Gamepad Inputs

`sampler2D gamepad` stores the status of gamepads connected to the PC.
The size of `gamepad` is `128x2`.
The status of buttons and axes are stored in `y = 0.0` and `y = 1.0`.

For example, `texture2D(gamepad, vec2(3. / 128., 0.))` returns 1.0 when the 3rd button is pressed.

See [examples](./examples/gamepad.frag) for actual usage.

### Loading images / videos

You can load images and videos by adding `IMPORTED` property in `.vedarc` or header comments.
If you write the path or URL of the resourece, it will be loaded automatically:

```glsl
/*
{
  "IMPORTED": {
    "image1": {
      "PATH": "1.jpg",
    },
    "image2": {
      "PATH": "../2.png",
    },
    "video1": {
      "PATH": "/Users/foo/Movies/1.mp4",
    },
    "video2": {
      "PATH": "http://example.com/2.mp4",
      "SPEED": 2,  // played 2x faster
    },
  },
}
*/
precision mediump float;
uniform vec2 resolution;

uniform sampler2D image1;
uniform sampler2D image2;
uniform sampler2D video1;
uniform sampler2D video2;

void main() {
	vec2 uv = gl_FragCoord.xy / resolution;

	gl_FragColor = (
		texture2D(image1, uv) +
		texture2D(image2, uv) +
		texture2D(video1, uv) +
    texture2D(video2, uv)
  );
}
```

The structure of `IMPORTED` properties is based on [Interactive Shader Format](https://www.interactiveshaderformat.com/).

See these examples for actual usage.

- [.vedarc](./examples/.vedarc)
- [image.frag](./examples/image.frag)
- [video.frag](./examples/video.frag)

### Multipass Rendering

VEDA supports multipass rendering.
You can define passes in `PASSES` property in `.vedarc` or header comments.

```glsl
/*
{
  "PASSES": [
    { "TARGET": "buffer" },
    {},
  ],
}
*/
```

The structure of `PASSES` property is based on [Interactive Shader Format](https://www.interactiveshaderformat.com/).
However, VEDA doesn't support `PERSISTENT` property.

VEDA supports `WIDTH` and `HEIGHT` in PASSES.
You can specify numbers for pixels or write expressions using `$WIDTH` and `$HEIGHT`.

```glsl
/*
{
  "PASSES": [
    {
      "TARGET": "buffer",
      "WIDTH": 512,  // 512px
      "HEIGHT": "$HEIGHT / 2",  // half of the render target (= Atom's width / pixelRatio)
    },
    {},
  ],
}
*/
```

See these examples for actual usage.

- [multipass.frag](./examples/multipass.frag)
- [multipass.vert](./examples/multipass.vert)

### Combining VS and FS

In `PASSES` you can specify vertex shader path from fragment shader, and vice versa.
For example, when you write header comments like below in fragment shader, VEDA will use `./vertex.vert` for vertex shader instead of default vertex shader.

```glsl
/*
{
  "PASSES": [{
    "vs": "./vertex.vert",
  }],
}
*/
```

See these examples for actual usage.

- [combination.frag](./examples/combination.frag) and [combination.vert](./examples/combination.vert)
- [particles.frag](./examples/particles.frag) and [particles.vert](./examples/particles.vert)

### Compute shader

You can write compute shaders using multipass rendering.
For compute shaders, we need to specify use float textures like this:

```glsl
/*{
  "PASSES": [
    {
      "fs": "./velocity.frag",
      "TARGET": "velocityTexture",
      "FLOAT": true,
    },
    {
      "fs": "./position.frag",
      "TARGET": "positionTexture",
      "FLOAT": true,
    },
    {
      "vs": "./scene.vert",
      "fs": "./scene.frag",
      "TARGET": "sceneTexture",
    },
    {}
  ]
}*/
```

To initialize textures, use `uniform int FRAMEINDEX`.

```glsl
uniform int FRAMEINDEX;

void main(){
    if (FRAMEINDEX == 0) {
        gl_FragColor = vec4(0);
    }
    else {
        // Do what you want
    }
}
```

See an [example](./examples/gpgpu/post.frag) for actual usage.

### Loading 3D models

You can load 3D models by passing file path to `MODEL` property in `PASSES`:

```glsl
/*{
  "PASSES": [{
    "vs": "./foo.vert",
    "MODEL": { "PATH": "./foo.obj" },
  }]
}*/
```

When you load `.obj` files in fragment shader, your shader is applied on the model defined in `.obj` file.
When you load `.obj` in vertex shader, you can use following attributes:

- `attribute vec3 position`
- `attribute vec3 normal`

Then

```glsl
precision mediump float;
attribute vec3 position;
attribute vec3 normal;
varying vec4 v_color;

void main(){
    gl_Position = vec4(position, 1);
    v_color = vec4(dot(normal, vec3(1)); // Lighting
}
```

If you use `.obj` files, you can also load `.mtl` files for materials:

```glsl
/*{
    PASSES: [{
        MODEL: {
            PATH: `foo.obj`,
            MATERIAL: `foo.mtl`,
        }
    }]
}*/
```

Materials are loaded as textures like `uniform sampler2D material0`, `uniform sampler2D material1`, etc.

See examples for more detail.

- `.obj` file ([frag](./examples/obj.frag), [vert](./examples/obj.vert))
- `.obj` with `.mtl` ([frag](./examples/obj-mtl.vert), [vert](./examples/obj-mtl.vert))
- `.json` file ([frag](./examples/json.frag), [vert](./examples/json.vert))

### glslify

VEDA supports [glslify](https://github.com/stackgl/glslify).

If `"glslify": true` is in the settings, VEDA bundles the code with glslify before evaluating.
Note that it will cause lint errors because linter-glsl doesn't support glslify.

See [examples](./examples/glslify.frag) for actual usage.

### Server Mode

If you wanna hide code and show only the shaders, you can use server mode.
When `server` is specified, VEDA launches a web server instead of running shaders in the background of Atom.

In this example, VEDA runs server on `http://localhost:3000`.
You can run shaders on the browsers by opening the url.

```glsl
/*
{
  "server": 3000,
}
*/
```

Warning: Currently we can't use videos/images outside the project directory in server mode.

See an [example](./examples/server.frag) for actual usage.

### Sound shader (experimental)

VEDA supports sound shaders like [Shadertoy](https://www.shadertoy.com/results?filter=soundoutput).

There are 2 command for sound shaders:

- `Veda: Load Sound Shader` (`alt-enter`): Play current shader as a sound shader.
- `Veda: Stop Sound Shader` (`alt-.`): Stop sound shaders.

In sound shader you have to define `vec2 mainSound(float time)` function instead of `void main()`.
`mainSound` takes current time stamp (`time`) and return the sample for stereo channels (`vec2`).

For example, this shader plays 440Hz and 660Hz sine wave in left and right channel.

```glsl
#define PI 3.141592653
vec2 mainSound(in float time) {
  return vec2(
    sin(2. * PI * time * 440.),
    sin(2. * PI * time * 660.)
  );
}
```

See an [example](./examples/sound.frag) for actual usage.

## Contributing

Any kind of issues or PRs are welcome!!😸

Before sending PRs, make sure you installed npm packages with `yarn`, not `npm`.
If not, we can't merge the PRs because this might cause errors and unnecessary diffs.

## Contributors

Thanks goes to these wonderful people ([emoji key](https://github.com/kentcdodds/all-contributors#emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->

<!-- prettier-ignore -->
| [<img src="https://avatars3.githubusercontent.com/u/1403842?v=4" width="100px;"/><br /><sub><b>Takayosi Amagi</b></sub>](https://gmork.in/)<br />[💬](#question-fand "Answering Questions") [💻](https://github.com/fand/veda/commits?author=fand "Code") [🎨](#design-fand "Design") [📖](https://github.com/fand/veda/commits?author=fand "Documentation") | [<img src="https://avatars0.githubusercontent.com/u/756562?v=4" width="100px;"/><br /><sub><b>Jonathan Giroux (Koltes)</b></sub>](http://koltes.digital)<br />[🐛](https://github.com/fand/veda/issues?q=author%3AKoltesDigital "Bug reports") [💻](https://github.com/fand/veda/commits?author=KoltesDigital "Code") [👀](#review-KoltesDigital "Reviewed Pull Requests") | [<img src="https://avatars3.githubusercontent.com/u/3462746?v=4" width="100px;"/><br /><sub><b>Cezary Kopias</b></sub>](http://cezary.kopias.pl)<br />[🐛](https://github.com/fand/veda/issues?q=author%3ACezaryKopias "Bug reports") [💡](#example-CezaryKopias "Examples") | [<img src="https://avatars1.githubusercontent.com/u/1937287?v=4" width="100px;"/><br /><sub><b>tanitta</b></sub>](http://tanitta.net/)<br />[💻](https://github.com/fand/veda/commits?author=tanitta "Code") [🤔](#ideas-tanitta "Ideas, Planning, & Feedback") | [<img src="https://avatars0.githubusercontent.com/u/630181?v=4" width="100px;"/><br /><sub><b>Yuya Fujiwara</b></sub>](http://ason.as)<br />[🐛](https://github.com/fand/veda/issues?q=author%3Aasonas "Bug reports") | [<img src="https://avatars0.githubusercontent.com/u/8775460?v=4" width="100px;"/><br /><sub><b>Rikuo Hasegawa</b></sub>](https://sp4ghet.com)<br />[🐛](https://github.com/fand/veda/issues?q=author%3Asp4ghet "Bug reports") [💻](https://github.com/fand/veda/commits?author=sp4ghet "Code") [💡](#example-sp4ghet "Examples") |
| :---: | :---: | :---: | :---: | :---: | :---: |

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification. Contributions of any kind welcome!

## Author

Takayosi Amagi

- Twitter: [@amagitakayosi](https://twitter.com/amagitakayosi/)
- GitHub: [fand](https://github.com/fand/)
- Blog: [マルシテイア](http://amagitakayosi.hatenablog.com/)

## License

MIT


CHANGELOG.md
# Changelog

## 2.14.3

- Fix errors on recording videos #246

## 2.14.2

- Fix errors on replacing textures
- Fix webcam example not working

## 2.14.0

- Support glslify in multipass shaders
- Add texture size to uniform as `vec2 [name]Size`
- Fix error on rebuild

## 2.13.2

- Fix OSC loader bug with masseges includes multiple slashes in the address

## 2.13.0

- Fix GIF loader broken by chromium update

## 2.12.2

- Fix materials on multiple models not loaded correctly

## 2.12.1

- Fix video textures not playing

## 2.12.0

- Show correct error positions on shaders using glslify
- Support `#include` macro
- Install glslangValidator automatically using [glslang-validator-prebuilt](https://github.com/fand/glslang-validator-prebuilt)

## 2.11.4

- fix: Update shader correctly on server mode (#126)

## 2.11.3

- Fix timing issue #124

## 2.11.1

- Rename recording commands / paths of recorded files

## 2.11.0

- Add commands to record videos

## 2.10.2

- Fixed bugs of file path conversion in server mode #94

## 2.10.0

- Support [glslify-import](https://github.com/glslify/glslify-import) #93

## 2.9.1

- Accept OSC bundle messages

## 2.9.0

- Rename `OBJ` to `MODEL`
- Suport `.json` file
- Suport `.mtl` file
- Add `MODEL.BLEND` to specify color blend mode

## 2.8.0

- Add `Veda: Toggle Fullscreen` command (`ctrl-escape`)

## 2.7.0

- Support `.obj` file

## 2.6.0

- Moved from Flowtype to TypeScript

## 2.5.2

- Support `WIDTH` `HEIGHT` in `PASSES`

## 2.5.0

- Support float textures with `FLOAT` in `PASSES`

## 2.4.2

- Fix comment bug #61

## 2.4.1

- Fix server mode not working #56

## 2.4.0

- Sound texture for sound shader

## 2.3.0

- Sound shader (experimental)

## 2.2.1

- Disable culling in vertex shader

## 2.2.0

- Add `fftSize` and `fftSmoothingTimeConstant` to settings

## 2.1.5

- Fix styles
- Allow loading cross-origin textures

## 2.1.4

- Rename `.liverc` to `.vedarc`

## 2.1.3

- Remove native addons from dependencies

## 2.1.2

- Fix style bug in video, camera etc.

## 2.1.1

- Fix OSC not working because of native modules

## 2.1.0

- Add `IMPORTED.SPEED` to Change video speed

## 2.0.1

- Fix OSC initialization on server mode

## 2.0.0

- OSC Support!

## 1.1.2

- Fix background-color of canvas

## 1.1.1

- Fix MIDI not working on Windows

## 1.1.0

- Support animated GIF

## 1.0.1

- Fix keybinding confliction on non-GLSL files

## 1.0.0

🔥🔥 VEDA Released 🔥🔥

- Rename package to `veda`
- Move GLSL runtime engine to [fand/veda](https://github.com/fand/veda)

## 0.13.1

- Fix server mode on Windows

## 0.13.0

- Add server mode

## 0.12.0

- Supports combining vs and fs

## 0.11.0

- Supports Multipass Rendering!

## 0.10.3

- Fix bugs on initializing configs

## 0.10.2

- Fix bugs on initializing vertex shaders

## 0.10.1

- Fix bugs in gamepad support and some errors

## 0.10.0

- Glslify Support!!

## 0.9.3, 0.9.4

- Fix bugs in `glsl-livecoder:toggle` #19

## 0.9.2

- Support `.fs` and `.vs` files

## 0.9.1

- Fix bugs around IMPORTED

## 0.9.0

- Support header comments to change settings per file

## 0.8.1

- Fix error on empty editor #14

## 0.8.0

- Vertex Shaders Support!!

## 0.7.2

- Fix error on glsl-livecoder:toggle #13

## 0.7.1

- Fix audio inputs not working
- Improve performance on disabling glsl-livecoder

## 0.7.0

- Keyboard Support!

## 0.6.0

- Gamepad Support!

## 0.5.0

- Webcam Support!

## 0.4.6

- Fix midi implementation

## 0.4.5

- Improve performance on loading images / videos
- Fix styles

## 0.4.3, 0.4.4

- Update README

## 0.4.2

- Fix unexpected sound output caused by audio inputs and videos

## 0.4.1

- Windows Support!!

## 0.4.0

- Support loading images / videos with `.liverc`

## 0.3.0

- Add `Pixel Ratio` and `Frameskip` to config
- Ignore non-GLSL files while watching active tab

## 0.2.0

- Add `uniform sampler2D note` to get MIDI notes

## 0.1.2

- Fix the style of the selected item in sidebar

## 0.1.1

- Revise README.md

## 0.1.0 - First Release

- Hello GLSL :zap:


src/index.ts
import { VedaStatus } from './constants';
import Wrapper from './wrapper';

let wrapper: Wrapper | null = null;

module.exports = {
    config: {
        glslangValidatorPath: {
            title: 'Custom glslangValidator path (optional)',
            description:
                'VEDA uses prebuilt glslangValidator by default. Specify custom path if you want.',
            type: 'string',
            default: '',
            order: 1,
        },
        pixelRatio: {
            title: 'Pixel Ratio',
            description:
                'The ratio of pixel per rendering call. Increasing pixel ratio will reduce CPU/GPU load.',
            type: 'number',
            default: 2,
            minimum: 0.5,
            maximum: 8,
            order: 2,
        },
        frameskip: {
            title: 'Frameskip',
            description:
                'Increasing frameskip will reduce CPU/GPU load. Default is 2 (= 30 fps).',
            type: 'integer',
            default: 2,
            minimum: 1,
            maximum: 10,
            order: 3,
        },
        vertexCount: {
            title: 'Vertex Count',
            description: 'The number of vertices in vertex shaders.',
            type: 'integer',
            default: 3000,
            minimum: 10,
            maximum: 20000,
            order: 4,
        },
        vertexMode: {
            title: 'Vertex Mode',
            description: 'How to draw vertices.',
            type: 'string',
            default: 'LINE_STRIP',
            enum: [
                { value: 'POINTS', description: 'POINTS' },
                { value: 'LINE_STRIP', description: 'LINE_STRIP' },
                { value: 'LINE_LOOP', description: 'LINE_LOOP' },
                { value: 'LINES', description: 'LINES' },
                { value: 'TRI_STRIP', description: 'TRI_STRIP' },
                { value: 'TRI_FAN', description: 'TRI_FAN' },
                { value: 'TRIANGLES', description: 'TRIANGLES' },
            ],
            order: 5,
        },
        fftSize: {
            title: 'FFT size',
            description:
                'Represents the size of the FFT to be used to determine the frequency domain.',
            type: 'integer',
            default: 2048,
            enum: [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768],
            order: 6,
        },
        fftSmoothingTimeConstant: {
            title: 'FFT smoothingTimeConstant',
            description:
                'Represents the averaging constant with the last analysis frame.',
            type: 'number',
            default: 0.8,
            minimum: 0,
            maximum: 1,
            order: 7,
        },
        recordingMode: {
            title: 'Recording Mode',
            description: 'File format to record images',
            type: 'string',
            default: 'mp4',
            enum: ['mp4', 'gif'],
            order: 8,
        },
    },

    activate(state: VedaStatus): void {
        require('atom-package-deps')
            .install('veda')
            .then((): void => this._activate(state));
    },

    _activate(state: VedaStatus): void {
        wrapper = new Wrapper(state);
    },

    deactivate(): void {
        wrapper && wrapper.destroy();
    },
};
